src/Service/CustomerInteractionService.php

```diff
@@ -0,0 +1,190 @@
+ *
+ * @author Conduction.nl <info@conduction.nl>, Robert Zondervan <robert@conduction.nl>
+ *
+ * @license EUPL-1.2 https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
+ */
+
+namespace CommonGateway\CustomerInteractionBundle\Service;
+
+use App\Entity\Gateway as Source;
+use CommonGateway\CoreBundle\Service\CallService;
+use Doctrine\ORM\EntityManagerInterface;
+use Symfony\Component\HttpFoundation\Response;
+
+class CustomerInteractionService
+{
+
+    /**
+     * The action configuration.
+     *
+     * @var array
+     */
+    private array $configuration;
+
+    /**
+     * The call service.
+     *
+     * @var CallService
+     */
+    private CallService $callService;
+
+    /**
+     * The entity manager.
+     *
+     * @var EntityManagerInterface
+     */
+    private EntityManagerInterface $entityManager;
+
+
+    public function __construct(
+        CallService $callService,
+        EntityManagerInterface $entityManager
+    ) {
+        $this->callService   = $callService;
+        $this->entityManager = $entityManager;
+
+    }//end __construct()
+
+
+    /**
+     * Find a source for a given url.
+     *
+     * @param string $url      The url of the object to be found.
+     * @param string $endpoint The resulting endpoint (the remainder of the path)
+     *
+     * @return Source|null The resulting source.
+     *
+     * @throws \Safe\Exceptions\UrlException
+     */
+    private function getSource(string $url, ?string &$endpoint): ?Source
+    {
+        // 1. Get the domain from the url
+        $parse    = \Safe\parse_url($url);
+        $location = $parse['scheme'].'://'.$parse['host'];
+
+        // 2.c Try to establish a source for the domain
+        $source = $this->entityManager->getRepository('App:Gateway')->findOneBy(['location' => $location]);
+
+        // 2.b The source might be on a path e.g. /v1 so if whe cant find a source let try to cycle
+        if ($source instanceof Source === false && isset($parse['path']) === true) {
+            foreach (explode('/', $parse['path']) as $pathPart) {
+                if ($pathPart !== '') {
+                    $location = $location.'/'.$pathPart;
+                }
+
+                $source = $this->entityManager->getRepository('App:Gateway')->findOneBy(['location' => $location]);
+                if ($source !== null) {
+                    $endpoint = str_replace($location, '', $url);
+                    break;
+                }
+            }
+        }
+
+        if ($source instanceof Source === false) {
+            return null;
+        }
+
+        return $source;
+
+    }//end getSource()
+
+
+    /**
+     * Extend an identificator object with the subobject.
+     *
+     * @param array $identificator The identificator object.
+     *
+     * @return array The resulting identificator object.
+     *
+     * @throws \Exception
+     */
+    public function extendIdentificator(array $identificator): array
```

Comment:
Can be a private function

Reply:
Here we might have a disagreement: I have always been taught that if there is no compelling reason to make a function private, it should be public (so, if there is a reason that the function really shouldn't be used outside of the function), but I don't see a compelling reason for that here.

Reply:
I have been thought the other way around, always make private unless you are sure it will be re-used outside your code or if you made it to be used outside even if it isn't yet